### ISSUE TITLE
Need to enforce tx.time < 500,000,000

### ISSUE THREAD

**Issue by gandrewstone** [mr-zwets/upgraded-SHA-gate#1]
https://github.com/mr-zwets/upgraded-SHA-gate/blob/6e9b98af225218deb796dab26ed14e27916c68c9/sha-gate-improved.cash#L75

Otherwise, an attacker could use nLockTime in its "seconds since the epoch" mode to cause the voting time to be ~30 min.  This would be bad because of "require(int(voteCountBytes) >= 0);".  So a voting period with no (or very few) votes looks like it would be ok.  The attacker's likelihood of success is proportional to the likelihood of no blocks within 30 min and the attacker's own hash percentage.

Caveat:  I am a raw BCH assembly coder, not cashscript.  Maybe cashscript automagically takes care of this for you?  If so just close the issue and sorry to waste your time.

**Comment by mr-zwets** [comment]
Thanks for bringing up the issue as I had simply not thought about this. Great that someone else is thinking about these contract details!!

The solution should be tweaked a bit to `require(tx.locktime < 500000000);`, the native introspection variable reading the nlocktime. This is because tx.time simply sets a nlocktime so can only restrict with a greater than sign ([as per the cashscript documentation](https://cashscript.org/docs/language/globals/#txtime)).

In my own words I would describe the issue as 
`make sure that the new value for startPeriodBytes is also expressed in blockheight and not UNIX time`

Will update this in the contract shortly, thanks again

**Comment by mr-zwets** [comment]
Thinking about this for a bit longer the current issue is not that the locktime can be set above 500,000,000 because the tx.time uses OP_CHECKLOCKTIMEVERIFY which requires that the argument and nlocktime are either both above the threshold or neither are. (clarified the comments on that section of the contract)

The drawback of the current implementation is that the new value for startPeriodBytes can start trailing behind when setting the locktime directly as the previous startPeriod + 2088. If instead of setting the locktime directly, the introspection variable for the nlocktime should be used and required to be greater than startPeriod + 2088 then this indeed allows for a nlocktime to be set above 500,000,000 and the above solution would apply

**Comment by gandrewstone** [comment]
Yes, but is there still an initial problem?  Since I am unfamiliar with cashscript, its not entirely clear to me how the contract is instantiated.  

Perhaps there's only ever one that creates a child, "baton" style?

Anyway, I had assumed creation was somehow under operator control via a multi-sig.    In that case 3-of-5 operators could conspire to take funds.  And also since having these operators is not actually decentralized (its distributed -- and is vulnerable to DoS attacks like we just saw with SmartBCH itself) I thought the intention would be to remove the operator role eventually, allowing anyone to propose withdrawals.

**Comment by mr-zwets** [comment]
closing this issue because as mentioned above `OP_CHECKLOCKTIMEVERIFY ` will require a transaction-level locktime value interpreted as blocks (not as timestamp) as the variable `int(startPeriodBytes) + 2088` will be below 500 million

there is no problem with the initial creation, all contracts with constructor arguments always need to be created in a sensible way. there is no need for a 'baton' or multisig.
